File: src/core/time/GameClock.ts
/**
 * Simulation calendar: ticks are the single source of truth; day/night,
 * seasons and years are all derived (see docs/TDD.md §4.3).
 *
 * The clock is bus-agnostic: `advance()` returns the calendar transitions
 * that occurred, and the owner (Simulation) publishes them as events.
 */
export const SIM_DT_MS = 100; // simulated milliseconds of real time per tick at speed x1
export const TICKS_PER_DAY = 240; // 24 s per in-game day at speed x1
export const DAYS_PER_SEASON = 12;
export const SEASONS = ["spring", "summer", "autumn", "winter"] as const;
export type Season = (typeof SEASONS)[number];
export const DAYS_PER_YEAR = DAYS_PER_SEASON * SEASONS.length;

export type ClockTransition =
  | { kind: "dayStarted"; day: number }
  | { kind: "seasonChanged"; season: Season; year: number }
  | { kind: "yearStarted"; year: number };

export class GameClock {
  tick = 0;

  /** Fraction of the current day in [0, 1): 0 = midnight, 0.5 = noon. */
  get timeOfDay(): number {
    return (this.tick % TICKS_PER_DAY) / TICKS_PER_DAY;
  }

  /** Absolute day count since world creation. */
  get day(): number {
    return Math.floor(this.tick / TICKS_PER_DAY);
  }

  /** Day within the current season, starting at 0. */
  get dayOfSeason(): number {
    return this.day % DAYS_PER_SEASON;
  }

  get season(): Season {
    const index = Math.floor(this.day / DAYS_PER_SEASON) % SEASONS.length;
    return SEASONS[index]!;
  }

  get year(): number {
    return Math.floor(this.day / DAYS_PER_YEAR);
  }

  /** Sunlight factor in [0, 1]: 0 at midnight, 1 at noon (cosine curve). */
  get daylight(): number {
    return 0.5 - 0.5 * Math.cos(this.timeOfDay * 2 * Math.PI);
  }

  /** Advance one tick and report any calendar transitions that occurred. */
  advance(): ClockTransition[] {
    const previousDay = this.day;
    const previousSeason = this.season;
    const previousYear = this.year;
    this.tick++;

    if (this.day === previousDay) return [];
    const transitions: ClockTransition[] = [{ kind: "dayStarted", day: this.day }];
    if (this.season !== previousSeason) {
      transitions.push({ kind: "seasonChanged", season: this.season, year: this.year });
    }
    if (this.year !== previousYear) {
      transitions.push({ kind: "yearStarted", year: this.year });
    }
    return transitions;
  }
}

File: unity/ImG.Core/Time/GameClock.cs
using System.Collections.Generic;

namespace ImG.Core.Time
{
    public enum Season { Spring = 0, Summer = 1, Autumn = 2, Winter = 3 }

    public enum TransitionKind { DayStarted, SeasonChanged, YearStarted }

    public readonly struct ClockTransition
    {
        public readonly TransitionKind Kind;
        public readonly int Day;
        public readonly Season Season;
        public readonly int Year;

        public ClockTransition(TransitionKind kind, int day, Season season, int year)
        {
            Kind = kind;
            Day = day;
            Season = season;
            Year = year;
        }
    }

    /// <summary>
    /// Calendrier de simulation : le tick est l'unique source de vérité ;
    /// jour/nuit, saisons et années en sont dérivés. Portage de
    /// <c>src/core/time/GameClock.ts</c>.
    /// </summary>
    public sealed class GameClock
    {
        public const int SimDtMs = 100;
        public const int TicksPerDay = 240;
        public const int DaysPerSeason = 12;
        public const int SeasonCount = 4;
        public const int DaysPerYear = DaysPerSeason * SeasonCount;

        public long Tick;

        /// <summary>Fraction du jour courant dans [0, 1) : 0 = minuit, 0.5 = midi.</summary>
        public double TimeOfDay => (double)(Tick % TicksPerDay) / TicksPerDay;

        /// <summary>Nombre de jours absolus depuis la création du monde.</summary>
        public long Day => Tick / TicksPerDay;

        public long DayOfSeason => Day % DaysPerSeason;

        public Season Season => (Season)((Day / DaysPerSeason) % SeasonCount);

        public long Year => Day / DaysPerYear;

        /// <summary>Facteur d'ensoleillement [0, 1] : 0 à minuit, 1 à midi (cosinus).</summary>
        public double Daylight => 0.5 - 0.5 * System.Math.Cos(TimeOfDay * 2 * System.Math.PI);

        /// <summary>Avance d'un tick et renvoie les transitions de calendrier survenues.</summary>
        public List<ClockTransition> Advance()
        {
            long previousDay = Day;
            Season previousSeason = Season;
            long previousYear = Year;
            Tick++;

            var transitions = new List<ClockTransition>();
            if (Day == previousDay) return transitions;

            transitions.Add(new ClockTransition(TransitionKind.DayStarted, (int)Day, Season, (int)Year));
            if (Season != previousSeason)
                transitions.Add(new ClockTransition(TransitionKind.SeasonChanged, (int)Day, Season, (int)Year));
            if (Year != previousYear)
                transitions.Add(new ClockTransition(TransitionKind.YearStarted, (int)Day, Season, (int)Year));
            return transitions;
        }
    }
}
